components/container_id/container_id.c

```diff
@@ -1,101 +1,113 @@
-        if (pos) {
-            return pos + dd_targets[i].len;
-        }
-    }
-    return NULL;
+    len = (size_t)pmatch[0].rm_eo - pmatch[0].rm_so;
+    if (len < MIN_ID_LEN || len > MAX_ID_LEN) return 0;
+
+    memcpy(buf, line + pmatch[0].rm_so, len);
+    buf[len] = '\0';
+
+    return len;
 }
 
 void datadog_php_container_id(char *buf, const char *file) {
     FILE *fp;
-    char line[1024];
+    char task_buf[MAX_ID_LEN + 1];
+
+    if (buf == NULL) return;
 
     buf[0] = '\0';
+    task_buf[0] = '\0';
 
-    if (file == NULL || file[0] == '\0') {
-        return;
-    }
+    if (file == NULL || file[0] == '\0' || (fp = fopen(file, "r")) == NULL) return;
 
-    fp = fopen(file, "r");
-    if (fp == NULL) {
+    /* According to the in-code docs:
+     *
+     * "Note that the translate table must either have been initialized by
+     * 'regcomp', with a malloc'ed value, or set to NULL before calling
+     * 'regfree'."
+     * https://elixir.bootlin.com/glibc/latest/source/posix/regex.h#L535
+     *
+     * For this reason we zero-initialize all of the 'regex_t' patterns so that
+     * we can still call regfree() on them if regcomp() fails to initialize the
+     * pattern buffer.
+     */
+    regex_t line_regex = {0};
+    regex_t task_regex = {0};
+    regex_t container_regex = {0};
+    int l_res = regcomp(&line_regex, LINE_REGEX, REG_EXTENDED | REG_NOSUB);
+    int t_res = regcomp(&task_regex, TASK_REGEX, REG_EXTENDED);
+    int c_res = regcomp(&container_regex, CONTAINER_REGEX, REG_EXTENDED);
+    if (l_res != 0 || t_res != 0 || c_res != 0) {
+        regfree(&container_regex);
+        regfree(&task_regex);
+        regfree(&line_regex);
+        fclose(fp);
         return;
     }
 
+    char line[1024];
     while (!feof(fp)) {
-        if (fgets(line, sizeof line, fp) != NULL) {
-            char *pos = dd_find_target(line);
-            if (pos) {
-                dd_extract_id(buf, pos);
-                if (buf[0] != '\0') {
-                    fclose(fp);
-                    return;
-                }
-            }
+        if (fgets(line, sizeof line, fp) == NULL) continue;
+
+        /* Match a valid cgroup line. */
+        if (regexec(&line_regex, line, 0, NULL, 0) != 0) continue;
+
+        /* Once we find a task ID we can stop scanning the cgroup file since
+         * those take precedence over a standard container ID. In many examples
+         * such as Fargate 1.4+, the task ID is on the last line of the file
+         * after many lines of valid container IDs. It is for this reason that
+         * we must scan the file until the very end even after finding a valid
+         * container ID.
+         */
+        size_t task_len = dd_find_id_regex(&task_regex, task_buf, line);
+        if (task_buf[0] != '\0') {
+            memcpy(buf, task_buf, task_len + 1);
+            break;
+        }
+        if (buf[0] == '\0') {
+            dd_find_id_regex(&container_regex, buf, line);
```

Comment:
I would expect a `memcpy` here, like the case above.

Reply:
There isn't an error here, but there might be a cleaner way to write this. The reason there is no `memcpy` here is because the happy path of finding a regular container ID copies directly into the main buffer of the caller (`buf`) but the edge case of a task ID uses an intermediary buffer (`task_buf`) since we'll likely have a valid container ID before finding a task ID.

But I think we can avoid the intermediary buffer by returning a "found" status from `dd_find_id_regex`. I'll rework this a bit and see if you think it's a bit cleaner.